loki/transform/dependency_transform.py

```diff
@@ -126,27 +129,50 @@ def transform_module(self, module, **kwargs):
             module.name = self.derive_module_name(module.name)
 
         # Module imports only appear in the spec section
-        imports = FindNodes(Import).visit(module.spec)
-        self.rename_imports(module, imports=imports, **kwargs)
+        self.rename_imports(module, imports=module.imports, **kwargs)
 
     def transform_file(self, sourcefile, **kwargs):
         """
-        In 'module' mode perform module-wrapping for dependnecy injection.
+        In 'module' mode perform module-wrapping for dependency injection.
         """
-        role = kwargs.get('role')
+        items = kwargs.get('items')
+        role = kwargs.pop('role', None)
+        targets = kwargs.pop('targets', None)
+
+        if not role and items:
+            # We consider the sourcefile to be a "kernel" file if all items are kernels
+            if all(item.role == 'kernel' for item in items):
+                role = 'kernel'
+
+        if targets is None and items:
+            # We collect the targets for file/module-level imports from all items
+            targets = [target for item in items for target in item.targets]
+
         if role == 'kernel' and self.mode == 'module':
             self.module_wrap(sourcefile, **kwargs)
 
+        for module in sourcefile.modules:
+            # Recursion into contained modules using the sourcefile's "role"
+            self.transform_module(module, role=role, targets=targets, **kwargs)
+
+        if items:
+            # Recursion into all subroutine items in the current file
+            for item in items:
+                self.transform_subroutine(item.routine, item=item, role=item.role, targets=item.targets, **kwargs)
```

Comment:
[no action] This interwoven subtlety highlights that the `DependencyTransform` is actually deeply flawed and needs to be re-designed. As already discussed offline, like as a combination of several individual transformations, so thanks for keeping this functional! :pray: